pkg/go/graph/weighted_graph_builder.go

```diff
@@ -18,60 +21,180 @@ func NewWeightedAuthorizationModelGraphBuilder() *WeightedAuthorizationModelGrap
+		if !typeAndRelationExists(model, tuplesetType, computedRelation) {
+			return fmt.Errorf("%w: Model cannot be parsed. %s type does not have defined %s relation", ErrInvalidModel, tuplesetType, computedRelation)
 		}
 
-		// NOTE: because we use a multigraph, one edge can include multiple lines, so we need to add each line individually.
-		iterLines := nextEdge.Lines
-		for iterLines.Next() {
-			nextLine := iterLines.Line()
-			castedEdge, ok := nextLine.(*AuthorizationModelEdge)
-			if !ok {
-				return nil, fmt.Errorf("%w: could not cast %v to AuthorizationModelEdge", ErrBuildingGraph, nextLine)
-			}
-			castedFromNode, ok := castedEdge.From().(*AuthorizationModelNode)
-			if !ok {
-				return nil, fmt.Errorf("%w: could not cast %v to AuthorizationModelNode", ErrBuildingGraph, castedEdge.From())
-			}
-			castedToNode, ok := castedEdge.To().(*AuthorizationModelNode)
-			if !ok {
-				return nil, fmt.Errorf("%w: could not cast %v to AuthorizationModelNode", ErrBuildingGraph, castedEdge.To())
-			}
-			wb.AddEdge(castedFromNode.uniqueLabel, castedToNode.uniqueLabel, castedEdge.edgeType, castedEdge.tuplesetRelation, castedEdge.conditions)
+		validTTU = true
+		rewrittenNodeName := fmt.Sprintf("%s#%s", tuplesetType, computedRelation)
+		nodeSource := wg.GetOrAddNode(rewrittenNodeName, rewrittenNodeName, SpecificTypeAndRelation)
+		typeTuplesetRelation := fmt.Sprintf("%s#%s", typeDef.GetType(), tuplesetRelation)
+
+		if wg.HasEdge(parentNode, nodeSource, TTUEdge, typeTuplesetRelation) {
+			// we don't need to do any condition update, only de-dup the edge. In case of TTU
+			// the direct relation will have the conditions
+			// for example, in the case of
+			// type group
+			//   relations
+			// 		define rel1: [user] or rel1 from parent
+			//		define parent: [group, group with condX]
+			// In the graph we only have one TTU edge from the OR node to the group#rel1 node, but there are no conditions associated to it
+			// the conditions are associated to the edge from group#parent node to the group node. This direct edge has two conditions: none and condX
+			continue
 		}
+
+		// new edge from "xxx#admin" to "yyy#viewer" tuplesetRelation on "yyy#parent"
+		wg.UpsertEdge(parentNode, nodeSource, TTUEdge, typeTuplesetRelation, relatedType.GetCondition())
+	}
+	if !validTTU {
+		return fmt.Errorf("%w: Model cannot be parsed. No type and relation link exists for tupleset relation %s and computed relation %s", ErrInvalidModel, tuplesetRelation, computedRelation)
 	}
+	return nil
+}
 
-	err = wb.AssignWeights()
-	if err != nil {
-		return nil, err
+func (wgb *WeightedAuthorizationModelGraphBuilder) parseComputed(wg *WeightedAuthorizationModelGraph, parentNode *WeightedAuthorizationModelNode, typeDef *openfgav1.TypeDefinition, relation string) {
+	nodeType := RewriteEdge
+	// e.g. define x: y. Here y is the rewritten relation
+	rewrittenNodeName := fmt.Sprintf("%s#%s", typeDef.GetType(), relation)
```

Comment:
performance: direct concatenation will be much faster than fmt.Sprintf
```suggestion
	rewrittenNodeName := typeDef.GetType() + "#" + relation
```